// messaging-api-telegram/lib/TelegramClient.js
'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _extends = Object.assign || function (target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i];for (var key in source) {if (Object.prototype.hasOwnProperty.call(source, key)) {target[key] = source[key];}}}return target;};

var _axios = require('axios');var _axios2 = _interopRequireDefault(_axios);
var _axiosError = require('axios-error');var _axiosError2 = _interopRequireDefault(_axiosError);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}











function handleError(err) {var _err$response$data =
  err.response.data;const error_code = _err$response$data.error_code,description = _err$response$data.description;
  const msg = `Telegram API - ${error_code} ${description || ''}`; // eslint-disable-line camelcase
  throw new _axiosError2.default(msg, err);
}

class TelegramClient {





  constructor(token) {_initialiseProps.call(this);
    this._token = token;
    this._axios = _axios2.default.create({
      baseURL: `https://api.telegram.org/bot${token}/`,
      headers: {
        'Content-Type': 'application/json' } });


  }

  get axios() {
    return this._axios;
  }

  get accessToken() {
    return this._token;
  }




  /**
     * https://core.telegram.org/bots/api#getwebhookinfo
     */


  /**
         * https://core.telegram.org/bots/api#setwebhook
         */





  /**
             * https://core.telegram.org/bots/api#deletewebhook
             */


  /**
                 * https://core.telegram.org/bots/api#getme
                 */


  /**
                     * https://core.telegram.org/bots/api#getuserprofilephotos
                     */






  /**
                         * https://core.telegram.org/bots/api#getfile
                         */





  /**
                             * Get link for file. This is extension method of getFile()
                             */







  /**
                                 * https://core.telegram.org/bots/api#getchat
                                 */





  /**
                                     * https://core.telegram.org/bots/api#getchatmemberscount
                                     */





  /**
                                         * https://core.telegram.org/bots/api#getchatmemberscount
                                         */





  /**
                                             * https://core.telegram.org/bots/api#getchatmemberscount
                                             */






  /**
                                                 * https://core.telegram.org/bots/api#sendmessage
                                                 */







  /**
                                                     * https://core.telegram.org/bots/api#sendphoto
                                                     */







  /**
                                                         * https://core.telegram.org/bots/api#sendaudio
                                                         */







  /**
                                                             * https://core.telegram.org/bots/api#senddocument
                                                             */







  /**
                                                                 * https://core.telegram.org/bots/api#sendsticker
                                                                 */







  /**
                                                                     * https://core.telegram.org/bots/api#sendvideo
                                                                     */







  /**
                                                                         * https://core.telegram.org/bots/api#sendvoice
                                                                         */







  /**
                                                                             * https://core.telegram.org/bots/api#sendvideonote
                                                                             */
  // sendVideoNote = (chatId: string, videoNote: string) =>
  //   this._request('/sendVideoNote', {
  //     chat_id: chatId,
  //     video_note: videoNote,
  //   });

  /**
   * https://core.telegram.org/bots/api#sendmediagroup
   */







  /**
       * https://core.telegram.org/bots/api#sendlocation
       */












  /**
           * https://core.telegram.org/bots/api#editmessagelivelocation
           */










  /**
               * https://core.telegram.org/bots/api#stopmessagelivelocation
               */





  /**
                   * https://core.telegram.org/bots/api#sendvenue
                   */
























  /**
                       * https://core.telegram.org/bots/api#sendcontact
                       */















  /**
                           * https://core.telegram.org/bots/api#sendchataction
                           */






  /**
                                * https://core.telegram.org/bots/api#editmessagetext
                                */






  /**
                                     * https://core.telegram.org/bots/api#editmessagecaption
                                     */






  /**
                                          * https://core.telegram.org/bots/api#editmessagereplymarkup
                                          */






  /**
                                               * https://core.telegram.org/bots/api#deletemessage
                                               */






  /**
                                                    * https://core.telegram.org/bots/api#kickchatmember
                                                    */







  /**
                                                         * https://core.telegram.org/bots/api#unbanChatMember
                                                         */






  /**
                                                              * https://core.telegram.org/bots/api#restrictChatMember
                                                              */







  /**
                                                                   * https://core.telegram.org/bots/api#promoteChatMember
                                                                   */







  /**
                                                                        * https://core.telegram.org/bots/api#exportChatInviteLink
                                                                        */





  /**
                                                                             * https://core.telegram.org/bots/api#setChatPhoto
                                                                             */






  /**
                                                                                  * https://core.telegram.org/bots/api#deleteChatPhoto
                                                                                  */





  /**
                                                                                       * https://core.telegram.org/bots/api#setChatTitle
                                                                                       */






  /**
                                                                                            * https://core.telegram.org/bots/api#setChatDescription
                                                                                            */






  /**
                                                                                                 * https://core.telegram.org/bots/api#setchatstickerset
                                                                                                 */






  /**
                                                                                                      * https://core.telegram.org/bots/api#deletechatstickerset
                                                                                                      */





  /**
                                                                                                           * https://core.telegram.org/bots/api#pinChatMessage
                                                                                                           */







  /**
                                                                                                                * https://core.telegram.org/bots/api#unpinChatMessage
                                                                                                                */





  /**
                                                                                                                     * https://core.telegram.org/bots/api#leaveChat
                                                                                                                     */





  /**
                                                                                                                          * https://core.telegram.org/bots/api#getchatmemberscount
                                                                                                                          */













  /**
                                                                                                                                 * https://core.telegram.org/bots/api#sendinvoice
                                                                                                                                 */



















  /**
                                                                                                                                      * https://core.telegram.org/bots/api#answershippingquery
                                                                                                                                      */











  /**
                                                                                                                                           * https://core.telegram.org/bots/api#answerprecheckoutquery
                                                                                                                                           */











  /**
                                                                                                                                                * https://core.telegram.org/bots/api#answerinlinequery
                                                                                                                                                */











  /**
                                                                                                                                                     * https://core.telegram.org/bots/api#sendgame
                                                                                                                                                     */







  /**
                                                                                                                                                          * https://core.telegram.org/bots/api#setgamescore
                                                                                                                                                          */







  /**
                                                                                                                                                               * https://core.telegram.org/bots/api#getgamehighscores
                                                                                                                                                               */}exports.default = TelegramClient;TelegramClient.connect = token => new TelegramClient(token);var _initialiseProps = function _initialiseProps() {this._request = (url, data) => this._axios.post(url, data).then(res => res.data, handleError);this.getWebhookInfo = () => this._request('/getWebhookInfo');this.setWebhook = url => this._request('/setWebhook', { url });this.deleteWebhook = () => this._request('/deleteWebhook');this.getMe = () => this._request('/getMe');this.getUserProfilePhotos = (userId, options) => this._request('/getUserProfilePhotos', _extends({ user_id: userId }, options));this.getFile = fileId => this._request('/getFile', { file_id: fileId });this.getFileLink = fileId => this.getFile(fileId).then(response => `https://api.telegram.org/file/bot${this._token}/${response.result.file_path}`);this.getChat = chatId => this._request('/getChat', { chat_id: chatId });this.getChatAdministrators = chatId => this._request('/getChatAdministrators', { chat_id: chatId });this.getChatMembersCount = chatId => this._request('/getChatMembersCount', { chat_id: chatId });this.getChatMember = (chatId, userId) => this._request('/getChatMember', { chat_id: chatId, user_id: userId });this.sendMessage = (chatId, text, options) => this._request('/sendMessage', _extends({ chat_id: chatId, text }, options));this.sendPhoto = (chatId, photo, options) => this._request('/sendPhoto', _extends({ chat_id: chatId, photo }, options));this.sendAudio = (chatId, audio, options) => this._request('/sendAudio', _extends({ chat_id: chatId, audio }, options));this.sendDocument = (chatId, document, options) => this._request('/sendDocument', _extends({ chat_id: chatId, document }, options));this.sendSticker = (chatId, sticker, options) => this._request('/sendSticker', _extends({ chat_id: chatId, sticker }, options));this.sendVideo = (chatId, video, options) => this._request('/sendVideo', _extends({ chat_id: chatId, video }, options));this.sendVoice = (chatId, voice, options) => this._request('/sendVoice', _extends({ chat_id: chatId, voice }, options));this.sendMediaGroup = (chatId, media, options) => this._request('/sendMediaGroup', _extends({ chat_id: chatId, media }, options));this.sendLocation = (chatId, { latitude, longitude }, options) => this._request('/sendLocation', _extends({ chat_id: chatId, latitude, longitude }, options));this.editMessageLiveLocation = ({ latitude, longitude }, options) => this._request('/editMessageLiveLocation', _extends({ latitude, longitude }, options));this.stopMessageLiveLocation = identifier => this._request('/stopMessageLiveLocation', _extends({}, identifier));this.sendVenue = (chatId, { latitude, longitude, title, address }, options) => this._request('/sendVenue', _extends({ chat_id: chatId, latitude, longitude, title, address }, options));this.sendContact = (chatId, { phone_number, first_name }, options) => this._request('/sendContact', _extends({ chat_id: chatId, phone_number, first_name }, options));this.sendChatAction = (chatId, action) => this._request('/sendChatAction', { chat_id: chatId, action });this.editMessageText = (text, options) => this._request('/editMessageText', _extends({ text }, options));this.editMessageCaption = (caption, options) => this._request('/editMessageCaption', _extends({ caption }, options));this.editMessageReplyMarkup = (replyMarkup, options) => this._request('/editMessageReplyMarkup', _extends({ reply_markup: replyMarkup }, options));this.deleteMessage = (chatId, messageId) => this._request('/deleteMessage', { chat_id: chatId, message_id: messageId });this.kickChatMember = (chatId, userId, options) => this._request('/kickChatMember', _extends({ chat_id: chatId, user_id: userId }, options));this.unbanChatMember = (chatId, userId) => this._request('/unbanChatMember', { chat_id: chatId, user_id: userId });this.restrictChatMember = (chatId, userId, options) => this._request('/restrictChatMember', _extends({ chat_id: chatId, user_id: userId }, options));this.promoteChatMember = (chatId, userId, options) => this._request('/promoteChatMember', _extends({ chat_id: chatId, user_id: userId }, options));this.exportChatInviteLink = chatId => this._request('/exportChatInviteLink', { chat_id: chatId });this.setChatPhoto = (chatId, photo) => this._request('/setChatPhoto', { chat_id: chatId, photo });this.deleteChatPhoto = chatId => this._request('/deleteChatPhoto', { chat_id: chatId });this.setChatTitle = (chatId, title) => this._request('/setChatTitle', { chat_id: chatId, title });this.setChatDescription = (chatId, description) => this._request('/setChatDescription', { chat_id: chatId, description });this.setChatStickerSet = (chatId, stickerSetName) => this._request('/setChatStickerSet', { chat_id: chatId, sticker_set_name: stickerSetName });this.deleteChatStickerSet = chatId => this._request('/deleteChatStickerSet', { chat_id: chatId });this.pinChatMessage = (chatId, messageId, options) => this._request('/pinChatMessage', _extends({ chat_id: chatId, messsage_id: messageId }, options));this.unpinChatMessage = chatId => this._request('/unpinChatMessage', { chat_id: chatId });this.leaveChat = chatId => this._request('/leaveChat', { chat_id: chatId });this.forwardMessage = (chatId, fromChatId, messageId, options) => this._request('/forwardMessage', _extends({ chat_id: chatId, from_chat_id: fromChatId, message_id: messageId }, options));this.sendInvoice = (chatId, product, options) => this._request('/sendInvoice', _extends({ chat_id: chatId }, product, options));this.answerShippingQuery = (shippingQueryId, ok, options) => this._request('/answerShippingQuery', _extends({ shipping_query_id: shippingQueryId, ok }, options));this.answerPreCheckoutQuery = (preCheckoutQueryId, ok, options) => this._request('/answerPreCheckoutQuery', _extends({ pre_checkout_query_id: preCheckoutQueryId, ok }, options));this.answerInlineQuery = (inlineQueryId, results, options) => this._request('/answerInlineQuery', _extends({ inline_query_id: inlineQueryId, results }, options));this.sendGame = (chatId, gameShortName, options) => this._request('/sendGame', _extends({ chat_id: chatId, game_short_name: gameShortName }, options));this.setGameScore = (userId, score, options) => this._request('/setGameScore', _extends({ user_id: userId, score }, options));this.
  getGameHighScores = (userId, options) =>
  this._request('/getGameHighScores', _extends({
    user_id: userId },
  options));};
